components/script/dom/worker.rs

```diff
@@ -100,6 +103,10 @@ impl Worker {
                           data: StructuredCloneData) {
         let worker = address.to_temporary().root();
 
+        if worker.r().closing.get() {
```

Comment:
Gecko appears to differentiate between normal events in the worker thread's queue and what it calls a ControlRunnable. The run loop in their worker prioritizes the control events and handles at most one normal event on an iteration. We could implement something similar by adding a control channel and moving to a select model for the loop.